File: AiSD5_C3/exercise_1/Pascal/main.pas
program Silnia;

{ Deklaracje wyprzedzające funkcji i procedur }
function WczytajLiczbe: Int64; forward;
function SilniaRekurencyjna(n: Int64): Int64; forward;
function SilniaIteracyjna(n: Int64): Int64; forward;
procedure WyswietlWyniki(iteracyjna, rekurencyjna: Int64); forward;
procedure Kontroler; forward;

{ Funkcja do wczytywania i walidacji liczby wejściowej }
function WczytajLiczbe: Int64;
var
    n: Int64;
    kod: Integer;
    s: string;
begin
    repeat
        Write('Podaj nieujemną liczbę do obliczenia silni: ');
        ReadLn(s);
        Val(s, n, kod);
        
        if (kod <> 0) or (n < 0) then
            WriteLn('Proszę podać prawidłową liczbę nieujemną!');
            
    until (kod = 0) and (n >= 0);
    
    WczytajLiczbe := n;
end;

{ Obliczanie silni metodą rekurencyjną }
function SilniaRekurencyjna(n: Int64): Int64;
begin
    { Warunek bazowy: dla n = 0 zwracamy 1 }
    if n = 0 then
        SilniaRekurencyjna := 1
    else
        { Przypadek rekurencyjny: n! = n * (n-1)! }
        SilniaRekurencyjna := n * SilniaRekurencyjna(n - 1);
end;

{ Obliczanie silni metodą iteracyjną }
function SilniaIteracyjna(n: Int64): Int64;
var
    wynik: Int64;
    i: Int64;
begin
    { Inicjalizacja wyniku jako 1 (uwzględnia przypadek n = 0) }
    wynik := 1;
    
    { Mnożenie liczb od 1 do n }
    for i := 1 to n do
        wynik := wynik * i;
        
    SilniaIteracyjna := wynik;
end;

{ Wyświetlanie wyników z odpowiednim formatowaniem }
procedure WyswietlWyniki(iteracyjna, rekurencyjna: Int64);
begin
    WriteLn('Silnia iteracyjnie:   ', iteracyjna);
    WriteLn('Silnia rekurencyjnie: ', rekurencyjna);
end;

{ Procedura sterująca }
procedure Kontroler;
var
    n: Int64;
    wynikIteracyjny, wynikRekurencyjny: Int64;
begin
    { Wczytaj dane wejściowe }
    n := WczytajLiczbe;
    
    { Oblicz silnię obiema metodami }
    wynikIteracyjny := SilniaIteracyjna(n);
    wynikRekurencyjny := SilniaRekurencyjna(n);
    
    { Wyświetl wyniki }
    WyswietlWyniki(wynikIteracyjny, wynikRekurencyjny);
end;

{ Program główny }
begin
    Kontroler;
end.

File: AiSD5_C3/exercise_1/Python/main.py
def wczytaj_liczbe():
    while True:
        try:
            n = int(input('Podaj nieujemną liczbę do obliczenia silni: '))
            if n < 0:
                raise ValueError
            return n
        except ValueError:
            print('Proszę podać prawidłową liczbę nieujemną!')

def silnia_rekurencyjna(n):
    if n == 0:
        return 1
    else:
        return n * silnia_rekurencyjna(n - 1)

def silnia_iteracyjna(n):
    wynik = 1
    for i in range(1, n + 1):
        wynik *= i
    return wynik

def wyswietl_wyniki(iteracyjna, rekurencyjna):
    print(f'Silnia iteracyjnie:   {iteracyjna}')
    print(f'Silnia rekurencyjnie: {rekurencyjna}')

def kontroler():
    n = wczytaj_liczbe()
    wynik_iteracyjny = silnia_iteracyjna(n)
    wynik_rekurencyjny = silnia_rekurencyjna(n)
    wyswietl_wyniki(wynik_iteracyjny, wynik_rekurencyjny)

if __name__ == "__main__":
    kontroler()